File: bundle-hack/update_csv.py
#!/usr/bin/env python3
# The syntax str | None was added in 3.10
# Remove when UBI9 moves to python 3.10+
from __future__ import annotations

import argparse
import base64
import io
import json
import os
import sys
import shutil

from ruamel.yaml import YAML


yaml = YAML()
yaml.preserve_quotes = True


parser = argparse.ArgumentParser()
parser.add_argument("manifest_directory", type=str,
                    help="Path to operator manifest directory.")
parser.add_argument("version", type=str,
                    help="Version string of the product to update in the manifest.")
args = parser.parse_args()


def main() -> None:
    build_manifest_file_path = get_manifest_file_path_from_directory(args.manifest_directory)
    if not build_manifest_file_path:
        print(f"Failed to find manifest in {args.manifest_directory}")
        sys.exit(1)

    print(f"Success to find manifest in {build_manifest_file_path}")
    with open(build_manifest_file_path) as f:
        manifest = yaml.load(f)
        print(f"Successfully loaded {build_manifest_file_path} from build.")

    add_required_annotations(manifest)
    replace_version(manifest)
    replace_icon(manifest)
    replace_images(manifest)
    remove_related_images(manifest)
    write_manifest(manifest)
    print(f"Successfully updated CSV manifest for {args.version}.")



def get_manifest_file_path_from_directory(d: str) -> str|None:
    for filename in os.listdir(d):
        if filename.endswith('clusterserviceversion.yaml'):
            return os.path.join(d, filename)

def add_required_annotations(m: dict) -> None:
    """
    Adds required annotations to the CSV file content represented as a dictionary.
    Errors out if 'metadata' or 'annotations' does not exist.

    :param m: A dictionary representing the operator CSV manifest.
    """
    required_annotations = {
        "features.operators.openshift.io/disconnected": "true",
        "features.operators.openshift.io/fips-compliant": "true",
        "features.operators.openshift.io/proxy-aware": "false",
        "features.operators.openshift.io/tls-profiles": "false",
        "features.operators.openshift.io/token-auth-aws": "false",
        "features.operators.openshift.io/token-auth-azure": "false",
        "features.operators.openshift.io/token-auth-gcp": "false"
    }

    if 'metadata' not in m:
        sys.exit("Error: 'metadata' does not exist in the CSV content.")

    if 'annotations' not in m['metadata']:
        sys.exit("Error: 'annotations' does not exist within 'metadata' in the CSV content.")

    for key, value in required_annotations.items():
        m['metadata']['annotations'][key] = value

    print("Successfully added required annotations.")

def replace_version(m: dict) -> None:
    manifest_version = m['spec']['version']
    print(f"Updating version references from {manifest_version}",
          f"to {args.version} in manifest.")
    m['metadata']['name'] = m['metadata']['name'].replace(manifest_version, args.version)
    m['metadata']['annotations']['olm.skipRange'] = m['metadata']['annotations']['olm.skipRange'].replace(manifest_version, args.version)
    m['spec']['replaces'] = 'file-integrity-operator.v' + manifest_version
    m['spec']['version'] = args.version
    print(f"Successfully updated the operator version references from",
          f"{manifest_version} to {args.version} in manifest.")


def replace_icon(m: dict) -> None:
    """Replace the upstream icon with a Red Hat branded image.

    Perform an in-place update of the icon data on the manifest.

    manifest(dict): A dictionary representing the operator CSV manifest
    """
    icon_path = "icons/icon.png"
    with io.open(icon_path, "rb") as f:
        base64_encoded_icon = base64.b64encode(f.read())
    icons = [{"base64data": base64_encoded_icon.decode(), "mediatype": "image/png"}]
    m['spec']['icon'] = icons
    print(f"Successfully updated the operator image to use icon in {icon_path}.")


def replace_images(m: dict) -> None:
    # Get operator build image, which is available through the environment
    # variable as a JSON string. This means we don't have to query brew
    # directly for data about builds.

    FIO_IMAGE_PULLSPEC = "quay.io/redhat-user-workloads/ocp-isc-tenant/file-integrity-operator/file-integrity-operator@sha256:c04b0a064ec006fa14a5ddc481b1c99cd08c559ccd0e545d15d154c5c8958e11"

    # This is incredibly specific to how the File Integrity Operator CSV is
    # written. If something changes upstream, it could break this format,
    # likely resulting in a KeyError (if the dictionary we're looking for
    # doesn't exist), IndexError (if the number of items in the list doesn't
    # match our expectations), or TypeError (if we're trying to fish a key out
    # of a list).
    m['spec']['install']['spec']['deployments'][0]['spec']['template']['spec']['containers'][0]['image'] = FIO_IMAGE_PULLSPEC
    container_env = m['spec']['install']['spec']['deployments'][0]['spec']['template']['spec']['containers'][0]['env']
    for e in container_env:
        if e['name'] == "RELATED_IMAGE_OPERATOR":
            e['value'] = FIO_IMAGE_PULLSPEC
    print("Successfully updated the operator image to use downstream builds.")


def remove_related_images(m: dict) -> None:
    # Remove relatedImages entirely from the CSV. OSBS will look for container
    # images in the manifest and populate them as relatedImages when the bundle
    # image is built, so that we don't have to. See
    # https://osbs.readthedocs.io/en/latest/users.html#pullspec-locations for
    # more information on how OSBS does this.
    del m['spec']['relatedImages']
    print("Removed relatedImages from operator manifest.")


def write_manifest(m: dict) -> None:
    old_csv_filename = get_manifest_file_path_from_directory('manifests')
    if not old_csv_filename:
        print(f"Failed to find manifest in {args.manifest_directory}")
        sys.exit(1)

    new_csv_filename = os.path.join('manifests', f"file-integrity-operator.v{args.version}.clusterserviceversion.yaml")
    os.rename(old_csv_filename, new_csv_filename)
    print(f"Successfully moved {old_csv_filename} to {new_csv_filename}.")
    with open(new_csv_filename, 'w') as f:
        yaml.dump(m, f)
        print(f"Successfully wrote updated manifest to {new_csv_filename}.")


if __name__ == "__main__":
    main()

File: bundle-hack/update_csv.go
package main

import (
	"encoding/base64"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"strings"

	"gopkg.in/yaml.v3"
)

func readCSV(csvFilename string, csv *map[string]interface{}) {
	yamlFile, err := os.ReadFile(csvFilename)
	if err != nil {
		log.Fatal(fmt.Sprintf("Error: Failed to read file '%s'", csvFilename))
	}

	err = yaml.Unmarshal(yamlFile, csv)
	if err != nil {
		log.Fatal(fmt.Sprintf("Error: Failed to unmarshal yaml file '%s'", csvFilename))
	}
}

func replaceCSV(csvFilename string, outputCSVFilename string, csv map[string]interface{}) {
	err := os.Remove(csvFilename)
	if err != nil {
		log.Fatal(fmt.Sprintf("Error: Failed to remofe file '%s'", csvFilename))
	}

	f, err := os.Create(outputCSVFilename)
	if err != nil {
		log.Fatal(fmt.Sprintf("Error: Failed to create file '%s'", outputCSVFilename))
	}

	enc := yaml.NewEncoder(f)
	defer enc.Close()
	enc.SetIndent(2)

	err = enc.Encode(csv)
	if err != nil {
		log.Fatal("Error: Failed encode the CSV into yaml")
	}
}

func getInputCSVFilePath(dir string) string {
	filenames, err := os.ReadDir(dir)
	if err != nil {
		log.Fatal("Failed to find manifest dir")
	}

	for _, filename := range filenames {
		if strings.HasSuffix(filename.Name(), "clusterserviceversion.yaml") {
			return filepath.Join(dir, filename.Name())
		}
	}

	log.Fatal("Failed to find CSV file in manifest dir")
	return ""
}

func getOutputCSVFilePath(dir string, version string) string {
	return filepath.Join(dir, fmt.Sprintf("file-integrity-operator.v%s.clusterserviceversion.yaml", version))
}

func addRequiredAnnotations(csv map[string]interface{}) {
	requiredAnnotations := map[string]string{
		"features.operators.openshift.io/cnf":              "false",
		"features.operators.openshift.io/cni":              "false",
		"features.operators.openshift.io/csi":              "false",
		"features.operators.openshift.io/disconnected":     "true",
		"features.operators.openshift.io/fips-compliant":   "true",
		"features.operators.openshift.io/proxy-aware":      "false",
		"features.operators.openshift.io/tls-profiles":     "false",
		"features.operators.openshift.io/token-auth-aws":   "false",
		"features.operators.openshift.io/token-auth-azure": "false",
		"features.operators.openshift.io/token-auth-gcp":   "false",
	}

	annotations, ok := csv["metadata"].(map[string]interface{})["annotations"].(map[string]interface{})
	if !ok {
		log.Fatal("Error: 'annotations' does not exist within 'metadata' in the CSV content")
	}

	for key, value := range requiredAnnotations {
		annotations[key] = value
	}
	fmt.Println("Added required annotations")
}

func replaceVersion(oldVersion, newVersion string, csv map[string]interface{}) {
	spec, ok := csv["spec"].(map[string]interface{})
	metadata, ok := csv["metadata"].(map[string]interface{})
	if !ok {
		log.Fatal("Error: 'spec' does not exist in the CSV content")
	}

	fmt.Println(fmt.Sprintf("Updating version references from %s to %s", oldVersion, newVersion))

	spec["version"] = newVersion
	spec["replaces"] = "file-integrity-operator.v" + oldVersion

	metadata["name"] = strings.Replace(metadata["name"].(string), oldVersion, newVersion, 1)

	annotations := metadata["annotations"].(map[string]interface{})
	annotations["olm.skipRange"] = strings.Replace(annotations["olm.skipRange"].(string), oldVersion, newVersion, 1)

	fmt.Println(fmt.Sprintf("Updated version references from %s to %s", oldVersion, newVersion))
}

func replaceIcon(csv map[string]interface{}) {

	s, ok := csv["spec"]
	if !ok {
		log.Fatal("Error: 'spec' does not exist in the CSV content")
	}
	spec := s.(map[string]interface{})

	iconPath := "../bundle/icons/icon.png"
	iconData, err := os.ReadFile(iconPath)
	if err != nil {
		log.Fatal(fmt.Sprintf("Error: Failed to read icon file '%s'", iconPath))
	}
	icon := make(map[string]string)
	icon["base64data"] = base64.StdEncoding.EncodeToString(iconData)
	icon["media"] = "image/png"

	var icons = make([]map[string]string, 1)
	icons[0] = icon

	spec["icon"] = icons

	fmt.Println(fmt.Sprintf("Updated the operator image to use icon in %s", iconPath))
}

func recoverFromReplaceImages() {
	if r := recover(); r != nil {
		log.Fatal("Error: It was not possible to replace RELATED_IMAGE_OPERATOR")
	}
}

func replaceImages(csv map[string]interface{}) {
	defer recoverFromReplaceImages()

	// Konflux will automatically update the image sha based on the most
	// recent builds. We want to peel off the SHA and append it to the Red
	// Hat registry so that the bundle image will work when it's available
	// there.
	konfluxPullSpec := "quay.io/redhat-user-workloads/ocp-isc-tenant/file-integrity-operator-release@sha256:b350fbfa98e1b3199f791e360d90a34c0b3745dfc68425f3b593a5f0291b0169"
	delimiter := "@"
	parts := strings.Split(konfluxPullSpec, delimiter)
	if len(parts) > 2 {
		log.Fatalf("Error: Failed to safely determine image SHA from Konflux pull spec: %s", konfluxPullSpec)
	}
	imageSha := parts[1]
	registry := "registry.redhat.io/compliance/openshift-file-integrity-rhel8-operator"
	redHatPullSpec := registry + delimiter + imageSha

	env, ok := csv["spec"].(map[string]interface{})["install"].(map[string]interface{})["spec"].(map[string]interface{})["deployments"].([]interface{})[0].(map[string]interface{})["spec"].(map[string]interface{})["template"].(map[string]interface{})["spec"].(map[string]interface{})["containers"].([]interface{})[0].(map[string]interface{})["env"].([]interface{})
	if !ok {
		log.Fatal("Error: 'env' with RELATED_IMAGE_OPERATOR does not exist in the CSV content")
	}

	for _, item := range env {
		variable := item.(map[string]interface{})
		if variable["name"] == "RELATED_IMAGE_OPERATOR" {
			variable["value"] = redHatPullSpec
		}
	}

	containersMap := csv["spec"].(map[string]interface{})["install"].(map[string]interface{})["spec"].(map[string]interface{})["deployments"].([]interface{})[0].(map[string]interface{})["spec"].(map[string]interface{})["template"].(map[string]interface{})["spec"].(map[string]interface{})["containers"].([]interface{})[0].(map[string]interface{})
	containersMap["image"] = redHatPullSpec

	fmt.Println("Updated the deployment manifest to use downstream builds")
}

func removeRelated(csv map[string]interface{}) {
	spec, ok := csv["spec"].(map[string]interface{})
	if !ok {
		log.Fatal("Error: 'spec' does not exist in the CSV content")
	}

	delete(spec, "relatedImages")
	fmt.Println("Removed the operator from operator manifest")
}

func main() {
	var csv map[string]interface{}

	manifestsDir := os.Args[1]
	oldVersion := os.Args[2]
	newVersion := os.Args[3]

	csvFilename := getInputCSVFilePath(manifestsDir)
	fmt.Println(fmt.Sprintf("Found manifest in %s", csvFilename))

	readCSV(csvFilename, &csv)

	addRequiredAnnotations(csv)
	replaceVersion(oldVersion, newVersion, csv)
	replaceIcon(csv)
	replaceImages(csv)
	removeRelated(csv)

	outputCSVFilename := getOutputCSVFilePath(manifestsDir, newVersion)
	replaceCSV(csvFilename, outputCSVFilename, csv)
	fmt.Println(fmt.Sprintf("Replaced CSV manifest for %s", newVersion))
}
